Add user docs for pod priority and preemption

bsalamat opened the pull request:
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> For 1.8 Features: set Milestone to `1.8` and Base Branch to `release-1.8`
> ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
>

@kubernetes/sig-scheduling-pr-reviews @wojtek-t @davidopp

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes.github.io/5328)
<!-- Reviewable:end -->

ref\ https://github.com/kubernetes/kubernetes/issues/47604

gyliu513 commented on docs/concepts/configuration/pod-priority-preemption.md:
````diff
@@ -0,0 +1,177 @@
+__alpha__ feature. It can be enabled by a command-line flag:
+
+```
+--feature-gates=PodPriority=true
````
Can you please also specify for which componment shall I specify this parameter?  I think we should enable it for both scheduler and apiserver.

tengqm replied:
isn't it the default that feature-gates are shared across all master components?

wojtek-t replied:
No matter what it is, I agree it should be mentioned explicitly.

gyliu513 commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+
+The following sections provide more information about these steps.
+
+## Enable Priority and Preemption
```
Can you please also mention that we should enable admission controller plugin here?

tengqm replied:
correct, there has to be `--admisstion-control=...,Priority` command line option.

gyliu513 replied:
Yes, but we should always have `ResourceQuota` as last one.

tengqm replied:
Also need `--runtime-config=scheduling.k8s.io/v1alpha1=true`, right? This could have worked because we are enabling this alpha API by default in current code base. However, the convention was to have all alpha APIs disabled by default, IIRC.

gyliu513 replied:
Hmm, I think that we will not hit https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/priority/admission.go#L148 the featuregate of priority if does not enable the admission control? @bsalamat

gyliu513 replied:
@bsalamat can you please show me the detail where did we add it to the list of default admission controllers? I did not find the code where it was added, am I missing anything?

bsalamat replied:
In this PR: https://github.com/kubernetes/kubernetes/pull/49322/files
I am not sure if it was the right thing to do given the downgrade issue we are observing.

gyliu513 replied:
Ah, I see, thanks @bsalamat , I think that when downgrade, the downgrade script should have some logic to delete this admission control plugin.

bsalamat replied:
That's what I would expect as well, but I am not sure if the downgrade process actually removes them, otherwise #52226 shouldn't have happened.

tengqm commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+objects can have any 32-bit integer value smaller than or equal to 1 billion. Larger
+numbers are reserved for system use.
+
+PriorityClass also has two optional fields: `globaleDefault` and `description`.
```
globaleDefault -> globalDefault

davidopp replied:
For symmetry, I would suggest to say the names of the required fields (`name` and `value`).

tengqm commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+PriorityClass also has two optional fields: `globaleDefault` and `description`.
+`globalDefault` indicates that the value of this PriorityClass should be used for
+pods without a `PriorityClassName`. Only one PriorityClass with `globalDefault` 
+set to true can exists in the system. If there is no PriorityClass with `globalDefault`
```
exists -> exist

tengqm commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+all the specified requirements of the pod, the pod is determined infeasible. At this
+point preemption logic is triggered for the pending pod. Let's call the pending pod P.
+Preemption logic tries to find a node where removal of pods with lower priority than
+P helps schedule P. If such node is found, one or more lower priority pods will
```
such node ->such a node

davidopp replied:
s/helps schedule P/would enable P to schedule on that node/

tengqm commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+the node?"
+
+If the answer is no, that node will not be considered for preemption. If the pending
+pod has inter-pod affinity on one or more of those lower priority pods on the node, the
```
on one or more ->with one or more?

bsalamat replied:
changed "on" to "to".

tengqm commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+pods and scheduler will find the pending pod infeasible on the node. As a result,
+it will not try to preempt any pods on that node.
+Scheduler will try to find other nodes for preemption and could possibly find another
+one, but there is no guarantee that such a node will be found.
```
Better move line 153-154 to around line 108 ? This may not be a problem specific to this affinity scenario.

bsalamat replied:
I don't think these lines blends well with the text at line 108.

tengqm commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+equal or higher priority pods.
+
+#### Cross Node Preemption
+When considering a node N for preemption in order to schedule pending pod P,
```
schedule pending -> schedule the pending

wojtek-t commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+title: Pod Priority and Preemption
+---
+
+[Pods](/docs/user-guide/pods) in Kubernetes 1.8 and later can have priority. Priority
```
Actually, wasn't priority itself added in 1.7?

davidopp replied:
I think no -- https://github.com/kubernetes/kubernetes/pull/48377 was merged Jul 19.

wojtek-t commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+indicates the importance of a pod. When a pod cannot be scheduled, scheduler tries
+to preempt lower priority pods in order to make scheduling of the pending pod possible.
+
+* TOC
```
When I clicked view - something is not generating correctly for it.

wojtek-t commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+They have so much time to finish their work and exit. If they don't, they will be
+killed. This graceful termination period creates a time gap between the point that
+scheduler preempts pods until the pending pod (P) can be scheduled on the node (N).
+When there are multiple victims on node N, they may exit or get terminated at
```
Actually, I don't think this problem is reserved to situation where there is many victims - it's just more likely there.

But in general, even if there is one victim, once this is finished you have no guarantee that the first pod you will be processing, will be the preemptor. So someone else may be scheduled in that place.

bsalamat replied:
That's correct. I fixed the text.

wojtek-t commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+get scheduled for a while. This scenario can cause problems in various clusters, but
+is particularly problematic in clusters where many new pods are created all the time.
+
+We intend to address this problem in beta version of pod preemption. The solution
```
Let's be very explici here: "Fixing this is a Beta blocker".

@davidopp - do you agree?

bsalamat replied:
I agree that it is a beta blocker, but should we put it in a user doc?
I am going to write a separate design doc on it. I will put the milestones there.

bsalamat replied:
Race condition in fixing and commenting.
I changed it to we will address in beta.

wojtek-t commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+are preempted. Current preemption algorithm does not perform preemption of pods
+on nodes other than N, when considering N for preemption.
+
+We may consider adding cross node preemption in future versions if we find an
```
Let's be explicit: "Fixing this will NOT be Beta or GA blocker".

bsalamat replied:
Again, I am a bit hesitant to expose users to our release planning in this kind of docs. I added a sentence to say that we cannot promise anything.

wojtek-t commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+later by other pods, which removes the benefits of having the complex logic of 
+respecting inter-pod affinity to lower priority pods.
+
+Our recommended solution for this problem is to create inter-pod affinity towards
```
Let's be explicit: "Fixing this will NOT be Beta or GA blocker".

bsalamat replied:
Added a piece of text to say that we cannot promise that it will be fixed in Beta or GA.

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+be backward compatible.
+
+## PriorityClass
+A PriorityClass object defines a mapping from a PriorityClassName to the integer
```
A would say "PriorityClass is a non-namespaced object that defines..."

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+A PriorityClass object defines a mapping from a PriorityClassName to the integer
+value of the priority. The higher the value, the higher the priority. PriorityClass
+objects can have any 32-bit integer value smaller than or equal to 1 billion. Larger
+numbers are reserved for system use.
```
You should explain the "system use" stuff here, or at the end of this doc, at least briefly. It's OK to point them to the design doc but something brief here would be useful so they don't have to go to the design doc.

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+of those PriorityClass names in their spec.
+The following YAML is an example of a pod configuration that uses the PriorityClass
+created above. Priority admission controller checks the spec and resolves the
+priority of the pod to 1,000,000.
```
do "1000000" instead of "1,000,000" because in a lot of places outside the US "," is the same as a US "." in numbers.

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+**Note 2:** Addition of a PriorityClass with `globalDefault` set to true does not
+change priority of existing pods. The value of such PriorityClass will be used only
+for pods created after the PriorityClass is added.
+
```
Add something to this section about what happens if you delete a PriorityClass object. Also you should mention that if you submit a pod that has a PriorityClassName that doesn't have a corresponding PriorityClass, the pod will be rejected by the admission controller.

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+## Preemption
+When pods are created, they go to a queue and wait to be scheduled. Scheduler picks a pod
+from the queue and tries to schedule it on a node. If no node is found that satisfies
+all the specified requirements of the pod, the pod is determined infeasible. At this
```
s/requirements/requirements (predicates)/

davidopp replied:
also, "determined infeasible" isn't really necessary, I would join the two sentences like
"all the specified requirements (predicates) of the pod, preemption logic is triggered for the pending pod."

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+from the queue and tries to schedule it on a node. If no node is found that satisfies
+all the specified requirements of the pod, the pod is determined infeasible. At this
+point preemption logic is triggered for the pending pod. Let's call the pending pod P.
+Preemption logic tries to find a node where removal of pods with lower priority than
```
s/pods/one or more pods/
(just so people understand that we don't necessarily evict all lower-priority pods)

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+#### Cross Node Preemption
+When considering a node N for preemption in order to schedule pending pod P,
+P may become feasible on N only when pods on other nodes are preempted. For 
+example, if there is anti-affinity from existing lower priority pods in a zone
```
since anti-affinity is symmetric, "towards" is just confusing. I would rewrite from this sentence to the end of the paragraph to be a little clearer and provide a concrete example. Something like:  "For example, P may have zone anti-affinity with some currently-running, lower-priority pod Q. P may not be schedulable on Q's node even if it preempts Q, for example if P is larger than Q so preempting Q does not free up enough space on Q's node and P is not high-priority enough to preempt other pods on Q's node. But P might theoretically be able to schedule on some other node by preempting Q *and* some pod(s) on this other node (preempting Q removes the anti-affinity violation, and preempting pod(s) on this other node frees up space for P to schedule there). The current preemption algorithm does not detect and execute such preemptions; that is, when determining whether P can schedule onto N, it only considers preempting pods on N."

bsalamat replied:
Done

bsalamat commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+#### Starvation of Preempting Pod
+When pods are preempted, the victims get their
+[graceful termination period](https://kubernetes.io/docs/concepts/workloads/pods/pod/#termination-of-pods).
+They have so much time to finish their work and exit. If they don't, they will be
```
Done

bsalamat commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+The current implementation of preemption considers a node for preemption only when
+the answer to this question is positive: "If all the pods with lower priority than
+the pending pod are removed from the node, can the pending pod be scheduled on
+the node?"
```
Done

bsalamat commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,177 @@
+pod has inter-pod affinity on one or more of those lower priority pods on the node, the
+inter-pod affinity rule cannot be satisfied in the absence of the lower priority
+pods and scheduler will find the pending pod infeasible on the node. As a result,
+it will not try to preempt any pods on that node.
```
Isn't the current sentence more accurate? We are talking about preemption here and scheduling will be the next step which may or may not happen on this node.

bsalamat commented:
> I would delete the "Our recommended solution..." sentence; getting into solutions is too much detail for user guide. But put in what @wojtek-t said.

I actually think this is very important for users to know. It shows them how they can have effective affinity rules in a cluster where preemption is enabled.

bsalamat commented:
I guess I addressed all the comments.
I seriously hope that one day I will receive some divine power that allows me to find all the reviewers' comments on github! :-/

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,197 @@
+scheduler preempts pods until the pending pod (P) can be scheduled on the node (N).
+In the meantime, scheduler keeps scheduling other pending pods. When one or more victims
+exit or get terminated, scheduler may place other pending pods on the node if the pods are
+ahead of P in the scheduling queue. In such a case, it is likely that
```
IIUC "ahead of P in the scheduling queue" isn't really relevant. It's just whichever pod is next in the scheduling queue when the victim terminates. It might be ahead of P or behind P. I would change this sentence to "As victims exit or get terminated, scheduler continues trying to schedule all of the pods in the pending queue, and one or more of them may be considered and scheduled to N before the scheduler considers scheduling P on N."

bsalamat replied:
Other pods must be ahead of P when victim exits or get terminated, otherwise P will be scheduled first. Anyway, I used almost the same text as you suggested.

davidopp replied:
I think we're just using definitions of "ahead of P." To me, "ahead of" means something is to the left of something else in a queue. But this is a circular queue and when you say "ahead of" you really mean "processed before" -- it may actually be to the right. Anyway, the new wording LGTM.

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,197 @@
+preempt other pods on node N or another node to let P schedule. This scenario may 
+be repeated again for the second and subsequent rounds of preemption and P may not
+get scheduled for a while. This scenario can cause problems in various clusters, but
+is particularly problematic in clusters where many new pods are created all the time.
```
Maybe say "in clusters with a high preemption rate" rather than "in clusters where many new pods are created all the time" as I think it's more related to the former?

bsalamat replied:
High creation rate is more of a problem that preemption rate. I changed it to "in clusters with a high creation rate". When creation rate is high, even a single preemption may face this issue.

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,197 @@
+one, but there is no guarantee that such a node will be found.
+
+We may address this issue in future versions, but we don't have a clear plan and cannot
+promise that it will be fixed in Beta or GA. Part
```
I think @wojtek-t would feel better if after "GA" you say "(i.e. we will not consider it a blocker for Beta or GA)."

bsalamat replied:
Sure. Done :)

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,197 @@
+We may address this issue in future versions, but we don't have a clear plan and cannot
+promise that it will be fixed in Beta or GA. Part
+of the reason is that finding the set of lower priority pods that satisfy all
+inter-pod affinity/anti-affinity rules is computationally expensive and adds
```
should you remove "anti-affinity" here?

bsalamat replied:
Sure. Done

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,197 @@
+later by other pods, which removes the benefits of having the complex logic of 
+respecting inter-pod affinity to lower priority pods.
+
+Our recommended solution for this problem is to create inter-pod affinity towards
```
s/towards/only towards/

bsalamat replied:
Done

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,197 @@
+pods on N.
+
+We may consider adding cross node preemption in future versions if we find an
+algorithm with reasonable performance, but we cannot promise anything at this point.
```
I think @wojtek-t would feel better if you add at the end something like "(it will not be considered a blocker for Beta and GA)."

bsalamat replied:
Done.

steveperry-53 commented:
@bsalamat, The base branch for this PR should be release-1.8. Do you have any objections to changing the base branch? Thanks.

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,197 @@
+[Pods](/docs/user-guide/pods) in Kubernetes 1.8 and later can have priority. Priority
+indicates importance of a pod relative to other pods. When a pod cannot be scheduled, scheduler tries
+to preempt (evict) lower priority pods in order to make scheduling of the pending pod possible.
+Soon, priority will also affect out-of-resource eviction ordering on the node.
```
The [design doc](https://github.com/kubernetes/community/pull/946) seems to have been merged just two weeks ago, so I don't think it's in 1.8.

bsalamat replied:
@dashpole told me that this would be a tiny change that he could do in a day. That's why I am not so sure.

davidopp commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,197 @@
+
+
+**Note 1:** If you upgrade your existing cluster and enable this feature, the priority
+of your existing pods will be considered to be zero.
```
I missed "and enable this feature" so I think it's fine to leave as-is.

davidopp commented:
LGTM

BTW @bsalamat if you haven't done it before, changing the base branch to release-1.8 should just require selecting release-1.8 from the drop-down at the top of this Github page where it says "bsalamat  wants to merge 2 commits into ..." (the "..." is where the drop-down should be)

gyliu513 commented:
Related with https://github.com/kubernetes/kubernetes/issues/52226

bsalamat commented:
@steveperry-53 Just changed the base branch. Thanks for the reminder.

k82cn commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,209 @@
+
+**Note:** Alpha features should not be used in production systems! Alpha 
+features are more likely to have bugs and future changes to them are not guaranteed to
+be backward compatible.
```
> not guaranteed to be backward compatible.

Do we have doc on the scope of alpha, beta and GA?

tengqm replied:
em... maybe you are referring to this? https://kubernetes.io/docs/reference/deprecation-policy/

bsalamat replied:
@k82cn This doc tries to state the features/improvements planned for Beta, but we don't have any other doc at this point.

k82cn commented on docs/concepts/configuration/pod-priority-preemption.md:
```diff
@@ -0,0 +1,209 @@
+to the integer value of the priority. The higher the value, the higher the 
+priority. The value is
+specified in `value` field which is required. PriorityClass
+objects can have any 32-bit integer value smaller than or equal to 1 billion. Larger
```
if bigger than 1 billion, will reject or ignore?

bsalamat replied:
Priority admission controller rejects them.

steveperry-53 commented:
@bsalamat, I have some suggestions for the text. Could you check the **Allow edits by maintainers** box in the right column. Thanks.

bsalamat commented:
@steveperry-53 Done

steveperry-53 commented:
@bsalamat, See https://github.com/kubernetes/kubernetes.github.io/pull/5328/commits/12bcac7895d37f16975290c4d52fdaaa469a62f6 for suggested edits. At one point in the doc, I have a TODO that asks this question:

TODO: Revise this next example. I don’t understand the example with Node M. I took a stab at it below, but I don’t think I’ve gotten it right. I don’t see why if we start by considering N, we need a third Node M.

bsalamat commented:
@steveperry-53 Scheduler doesn't do cross node preemption in this version. So, when scheduler considers pod P to run on node N and pod P has anti-affinity to pod Q and pod Q is running on a different node, pod P will be deemed unschedulable on node N no matter how many pods are preempted on N.

steveperry-53 commented:
@bsalamat, Are you ready for me to merge this?

bsalamat commented:
@steveperry-53 Yes, I think this is ready now.

bsalamat commented:
@steveperry-53 Now that the PR is merged, where is its permanent location that I can link to in our release notes?
